### DISCUSSION TITLE
Modals with URLS

### DISCUSSION THREAD

**Discussion by NeilWkz** [TanStack/router#1966] [Q&A]
How do you do this with tanstack router? 

Our platform team recently migrated my company onto Tanstack; thanks to the community, it's been really nice to use so far.
I'll describe the behaviour that I've been tasked with executing: 
I'm doing an adaptive render that would render a full mobile page with a URL, and on desktop we render a full page with content in a modal overlay. The adaptive render part is no problem.
I know how i would execute the modal with a URL pattern with [react-router](https://v5.reactrouter.com/web/example/modal-gallery) but I'm struggling to work out from the docs how I would execute this with Tanstack. I know in Next this behaviour is called [Parallel Routes](https://nextjs.org/docs/app/building-your-application/routing/parallel-routes?eQSRP=vK8FJisoXu1D0JIT5kAN) and there is a documentation page with that name [here on tanstack](https://tanstack.com/router/latest/docs/framework/react/guide/parallel-routes?eQSRP=NqEthu2MElgsjyHPUAPW), but it has a TODO.

Does anyone have any examples of how to execute this pattern in Tanstack? Anything at all for a starting point would be gratefully recieved; sandboxes, links to other repos.

**Comment by NeilWkz** [comment]
> How do you do this with tanstack router? 
> 
> Our platform team recently migrated my company onto Tanstack; thanks to the community, it's been really nice to use so far.
> I'll describe the behaviour that I've been tasked with executing: 
> I'm doing an adaptive render that would render a full mobile page with a URL, and on desktop we render a full page with content in a modal overlay. The adaptive render part is no problem.
> I know how i would execute the modal with a URL pattern with [react-router](https://v5.reactrouter.com/web/example/modal-gallery) but I'm struggling to work out from the docs how I would execute this with Tanstack. I know in Next this behaviour is called [Parallel Routes](https://nextjs.org/docs/app/building-your-application/routing/parallel-routes?eQSRP=vK8FJisoXu1D0JIT5kAN) and there is a documentation page with that name [here on tanstack](https://tanstack.com/router/latest/docs/framework/react/guide/parallel-routes?eQSRP=NqEthu2MElgsjyHPUAPW), but it has a TODO.
> 
> Does anyone have any examples of how to execute this pattern in Tanstack? Anything at all for a starting point would be gratefully recieved; sandboxes, links to other repos.

No updates here so I'm just bumping to keep alive. If I get some examples I am happy to write some docs and open a pr

**Reply by schiller-manuel** [reply]
as parallel routes are not supported yet, you could use a query param and render the modal if that query param is set.

**Reply by NeilWkz** [reply]
Thanks for the reply.
 Yes that's a possible work around but not suitable for our usecase because these are full URLs on mobile and we want them to be crawlable for SEO purposes, so using a parameter would mean that crawlers would ignore it. 

Ideally we would be able to use a normal Tanstack Link (so it would be navigatable with javascript off for crawlers) and use a function to detect if we are on desktop and prevent default, and then open the modal and change the URL with something like `window.history.replaceState` and open the modal. The issue with that approach is that we cannot replace state without Tanstack navigating. I can't find any way to update the URL without also navigating. In RR I would call something like `navigate('/AdaptiveUrl', { replace: true })`

**Reply by ssatriya** [reply]
I wonder what's you solution to achieve this?

**Reply by NeilWkz** [reply]
In the end, I had to accept that until there is parallel route support we cannot support an adaptive render. We had to redesign our UI to go with simple static pages instead of modals that rendered only on desktop.

**Reply by ssatriya** [reply]
Why not go with the query param for the time being? I actually tried tanstack start and thought about using search param for now.

**Reply by NeilWkz** [reply]
Google won't crawl a query param as a separate page because the meta data in the head, canonical URL won't change when a query string changes. It wont be searchable. So it was not fit for our requirements. If you don't need the page to be searchable or it's not information with SEO value then it may work for you

**Reply by ssatriya** [reply]
Ah I see, thanks for the answer.

**Reply by tannerlinsley** [reply]
Google will crawl any link it finds unless you tell it not to (many ways to do this), including urls with query/search params. It’s important to use canonical urls (which can include search params) to identify unique pages/content. 

We plan on supporting parallel route soon and they will most definitely use search params to track the state of non primary slots.

**Reply by NeilWkz** [reply]
Thanks Tanner, glad to hear parallel route support is in the works.

Yes trade offs, Adding a mechanism to change the canonical URL for the querystring was more complicated for us so it was easier to comprise the ui. YMMV etc

**Reply by smilingkylan** [reply]
For the record I have had some issues with Tanstack Start and modals with state saved in the URL. Specifically, when I attempt to save form state into the URL, any change to a search param causes the modal to disappear the re-appear which is not desirable. I've really been having issue attempting to implement this
